src/io.rs

```diff
@@ -32,30 +39,323 @@ where
+        let len = match self.read_next_len_unchecked()? {
+            Some(len) => len,
+            None => return Ok(None),
+        };
+        if (buf.len() as u64) < len {
+            self.reader.seek(SeekFrom::Current(8 + len as i64))?;
+            return Err(RecordReadError::BufferTooShort { needed: len });
+        }
+        let mut len_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut len_bytes, len);
+        let len_ok = self.checksum(&mut len_bytes)?;
+
+        let slice = &mut buf[0..len as usize];
+        self.read_bytes_exact_unchecked(slice)?;
+        let bytes_ok = self.checksum(slice)?;
+
+        if bytes_ok && len_ok {
+            return Ok(Some(len as usize));
+        } else {
+            if !len_ok {
+                return Err(RecordReadError::CorruptFile);
+            }
+            return Err(RecordReadError::CorruptRecord);
+        }
+    }
+    /// Allocate a Vec<u8> on the heap and read the next record into it.
+    /// Returns the filled Vec, if successful.
+    /// Returns None, if it could read exactly 0 bytes (indicating EOF)
+    /// # Example
+    /// ```
+    /// use tensorflow::io::{RecordReadError, RecordReader, RecordWriter};
+    /// use std::{io::Cursor, rc::Rc};
+    /// let mut buf = Vec::new();
+    /// let mut rc = Rc::new(&mut buf);
+    /// let records = vec!["foo", "barr", "baz"];
+    /// {
+    ///     let mut writer = RecordWriter::new(Rc::get_mut(&mut rc).unwrap());
+    ///     for rec in records.iter() {
+    ///         writer.write_record(rec.as_bytes()).unwrap();
+    ///     }
+    /// }
+    /// let read = std::io::BufReader::new(Cursor::new(buf));
+    /// let mut reader = RecordReader::new(read);
+    /// let mut i = 0;
+    /// loop {
+    ///     let next = reader.read_next_owned();
+    ///     match next {
+    ///         Ok(res) => match res {
+    ///             Some(vec) => assert_eq!(&vec[..], records[i].as_bytes()),
+    ///             None => break,
+    ///         },
+    ///         Err(RecordReadError::CorruptFile) | Err(RecordReadError::IoError { .. }) => {
+    ///             break;
+    ///         }
+    ///         _ => {}
+    ///     }
+    ///     i += 1;
+    /// }
+    /// ```
+    pub fn read_next_owned(&mut self) -> Result<Option<Vec<u8>>, RecordReadError> {
+        let len = match self.read_next_len_unchecked()? {
+            Some(len) => len,
+            None => return Ok(None),
+        };
+        let mut vec = vec![0u8; len as usize];
+        let mut len_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut len_bytes, len);
+        let len_ok = self.checksum(&mut len_bytes)?;
+        self.read_bytes_exact_unchecked(&mut vec)?;
+        let bytes_ok = self.checksum(&vec)?;
+        if bytes_ok && len_ok {
+            return Ok(Some(vec));
+        } else {
+            if !len_ok {
+                return Err(RecordReadError::CorruptFile);
+            }
+            return Err(RecordReadError::CorruptRecord);
+        }
```

Comment:
This could be simplified to:

```
if !len_ok {
  Err(RecordReadError::CorruptFile)
} else if !bytes_ok {
  Err(RecordReadError::CorruptRecord)
} else {
  Ok(Some(len as usize))
}
```